src/Language/Fixpoint/Solver/PLE.hs

```diff
@@ -830,6 +857,81 @@ evalApp γ ctx e0 es et
+-- the standard PLE espansion, see the following example:
+-- We have a function f:
+-- define f (x : int) : int = {(x)}
+-- And we need to prove some constraint of this shape
+-- { f = something }
+-- The original PLE cannot do anything on f as it's not fully applied, instead
+-- if we perform eta expansion on f we obtain the following constraint
+-- { \y:Int -> f y = something }
+-- And now PLE can perform the unfolding of f as it's now fully saturated
+-- { \y:Int -> y = something }
+-- Clearly we need the higerorder flag active as we are generating lambda in
+-- the constraints.
+evalApp γ _ctx e0 es _et
+  | EVar f <- dropECst e0
+  , Just eq <- Map.lookup f (knAms γ)
+  = do
+    isHOAllowed <- gets hoFlag
+    if isHOAllowed then do
+      let expectedArgs = eqArgs eq
+      let nProvidedArgs = length es
+      let nArgsMissing = length expectedArgs - nProvidedArgs
+
+      -- Fresh names for the eta expansion
+      etaNames <- makeFreshEtaNames nArgsMissing
+      -- For some reason the application is expected to be already 
+      -- monomorphized, so we can't extract the sorts directly form
+      -- the equation
+      let etaArgsMono = drop nProvidedArgs <$> extractMonoSign e0
+      case etaArgsMono of
+          -- This case is pretty sus as we shouldnt generate funcitons that do not
+          -- have a type annotation
+          Nothing -> do
+            -- Hack to recover general mono signature, by looking directly
+            -- at the types in the equation
+            let etaArgsType = fmap snd $ drop nProvidedArgs $ eqArgs eq
+            if not $ any hasTyVar etaArgsType then do
+              performEtaExpansion e0 es etaNames etaArgsType
+            else do
+              -- There is at least a non monomorphized type variables
+              -- We cant conclude anything
+              pure (Nothing, noExpand)
+          Just etaArgsType -> do
+            performEtaExpansion e0 es etaNames etaArgsType
+    else do
+      pure (Nothing, noExpand)
+  where
+    hasTyVar :: Sort -> Bool
+    hasTyVar (FObj _) = True
+    hasTyVar (FApp l r) = hasTyVar l || hasTyVar r
+    hasTyVar (FFunc l r) = hasTyVar l || hasTyVar r
+    hasTyVar (FAbs _ i) = hasTyVar i
+    hasTyVar _ = False
+
+    extractMonoSign :: Expr -> Maybe [Sort]
+    extractMonoSign (ECst _ sort) = Just $ flatten sort
+    extractMonoSign _ = Nothing
+
+    flatten :: Sort -> [Sort]
+    flatten (FFunc t ts) = t : flatten ts
+    flatten t = [t]
+
+    performEtaExpansion :: Expr -> [Expr] -> [Symbol] -> [Sort] -> EvalST (Maybe Expr, FinalExpand)
+    performEtaExpansion fun actualArgs etaNames etaArgsType = do
+      let etaVars = zipWith (\name ty -> ECst (EVar name) ty) etaNames etaArgsType
+      let fullBody = eApps fun (actualArgs ++ etaVars)
+      let etaExpandedTerm = mkLams fullBody (zip etaNames etaArgsType)
+
+      modify $ \st ->
+        st { evNewEqualities = S.insert (eApps fun es, etaExpandedTerm) (evNewEqualities st) }
```

Comment:
Shouldn't this be skipped when `etaNames` is empty?

Reply:
It wont happen, eta name is empty only if the function is fully applied and that case is already covered by the first case of `evalApp`

Reply:
:+1: Good point. It would deserve adding a comment.